crates/bevy_tasks/src/priority_executor.rs

```diff
@@ -0,0 +1,195 @@
+use event_listener::Event;
+use futures_lite::FutureExt;
+
+/// Task priority.
+#[derive(Debug, Clone, Copy)]
+pub enum Priority {
+    FinishWithinFrame = 0,
+    AcrossFrame = 1,
+    IO = 2,
+}
+
+#[derive(Debug)]
+pub struct PriorityExecutor<'a> {
+    ex: [Executor<'a>; 3],
+    event: Event,
+}
+
+impl<'a> Default for PriorityExecutor<'a> {
+    fn default() -> Self {
+        PriorityExecutor::new()
+    }
+}
+
+impl<'a> PriorityExecutor<'a> {
+    /// Creates a new executor.
+    pub fn new() -> PriorityExecutor<'a> {
+        PriorityExecutor {
+            ex: [Executor::new(), Executor::new(), Executor::new()],
+            event: Event::new(),
+        }
+    }
+
+    pub fn spawn<T: Send + 'a>(
+        &self,
+        priority: Priority,
+        future: impl Future<Output = T> + Send + 'a,
+    ) -> Task<T> {
+        self.ex[priority as usize].spawn(future)
+    }
+
+    pub async fn run<T>(&self, future: impl Future<Output = T>) -> T {
+        let run_forever = async {
+            loop {
+                loop {
+                    if !self.ex[0].try_tick() {
+                        break;
+                    }
+                }
+
+                let t1 = self.ex[1].tick();
+                let t2 = self.ex[2].tick();
+                self.event.listen().or(t1).or(t2).await
+            }
+        };
+
+        future.or(run_forever).await
+    }
+
+    pub fn tick(&self) {
+        self.event.notify(usize::MAX);
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::{Arc, RwLock};
+    use std::thread;
+    use std::time::Duration;
+
+    use futures_lite::future;
+    use instant::Instant;
+
+    use super::*;
+
+    #[inline(never)]
+    async fn a_task(duration: Duration, priority: Priority, counters: Arc<RwLock<[usize; 3]>>) {
+        let start = Instant::now();
+        while Instant::now() - start < duration {}
+        counters.write().unwrap()[priority as usize] += 1;
+    }
+
+    #[test]
+    fn test_priority_executor() {
+        let tasks_completed: Arc<RwLock<[usize; 3]>> = Arc::new(RwLock::new([0, 0, 0]));
+        let mut tasks_dispatched: [usize; 3] = [0, 0, 0];
+        let executor = PriorityExecutor::new();
+        let executor = Arc::new(executor);
+        let (shutdown_tx, shutdown_rx) = async_channel::unbounded::<()>();
+
+        let num_cpus = num_cpus::get();
+        let mut job_handles = vec![];
+
+        for _ in 0..num_cpus {
+            let executor = executor.clone();
+            let shutdown_rx = shutdown_rx.clone();
+            // Spawn a thread running the executor forever.
+
+            let job_handle = thread::spawn(move || {
+                let shutdown_future = executor.run(shutdown_rx.recv());
+                // Use unwrap_err because we expect a Closed error
+                future::block_on(shutdown_future).unwrap_err();
+                //println!("worker thread {} exits", x)
+            });
+            job_handles.push(job_handle);
+        }
```

Comment:
```suggestion
        let job_handles =  0..num_cpus.map(|_| {
            let executor = executor.clone();
            let shutdown_rx = shutdown_rx.clone();
            // Spawn a thread running the executor forever.

            thread::spawn(move || {
                let shutdown_future = executor.run(shutdown_rx.recv());
                // Use unwrap_err because we expect a Closed error
                future::block_on(shutdown_future).unwrap_err();
                //println!("worker thread {} exits", x)
            })
        }).collect::<Vec<_>>();
```